### ISSUE TITLE
[$500] Status - Error message & emoji status remains when clear status timer has elapsed

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#33128]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.13-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/24620
## Action Performed:
Precondition: user should be Signed In
1. Open https://staging.new.expensify.com/
2. Click on Profile icon > Status
3. Set any status emoji and click on Save
4. Click on Status
5. Click on "Clear after" > Custom
6. Set the time 5 min ahead and click on Save button
7. Wait about 6 min or more
8. Open the Clear status option

## Expected Result:
The status should be cleared when the clear timer has elapsed

## Actual Result:
Please choose a time at least one minute ahead" error message is displayed, and the emoji status remains when the clear status timer has elapsed

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>  

https://github.com/Expensify/App/assets/78819774/7644a577-f91c-42b3-b93c-9b7b310f3e82

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013df2712d0a3d5327</li>
        <li>Upwork Job ID: 1737192168446410752</li>
        <li>Last Price Increase: 2024-01-09</li>
    </ul>
</details>

**Comment by stitesExpensify** [comment]
Looking into this

**Comment by abzokhattab** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
status is not saved to the backend 
### What is the root cause of that problem?
We don’t write the status data to the backend on changing the status

function updateDraftCustomStatus(status) {
    Onyx.merge(ONYXKEYS.CUSTOM_STATUS_DRAFT, status);
}
 
### What changes do you think we should make in order to solve the problem?
add `updateCustomStatus(status)` 

```js
function updateDraftCustomStatus(status) {
    Onyx.merge(ONYXKEYS.CUSTOM_STATUS_DRAFT, status);
    updateCustomStatus(status);
}

```

**Comment by stitesExpensify** [comment]
@abzokhattab have you tested that? I don't think that is the correct cause. We don't need to set the status when we are just creating a draft, and when you actually click save it looks like the request is working properly.

When you refresh the page after your timer, the status disappears. That leads me to believe that the issue is that we are not pushing the updates to all users properly.


IMO this is not a blocker even though it is genuinely broken. The Status feature that we merged was massive, and reverting it is not feasible. I will work on a fix in web-e first thing next week

**Comment by abzokhattab** [comment]
@stitesExpensify , I found out the comment wasn't working, so I hid it. After testing, I think the problem is related to time zones. I tried using a date from 6 hours ago, but the status didn't update even after that time had passed. However, when I checked it just now, the status was updated. It seems there's a mismatch in how the time zone is interpreted. I suspect the server, where the backend is running, is in the USA, causing the entered date to be parsed as a US date. This might be delaying the job trigger by a few hours. My suggestion is to send the date from the frontend as UTC, so the backend can parse it correctly and handle it accordingly.

**Comment by abzokhattab** [comment]
Just tried to test it again .. this time the status is not cleared 😂 even after almost 11 hours of the custom date .. so i guess the backend doesnt clear it.

**Comment by abzokhattab** [comment]
If we have decided to tackle the status clearing process in the frontend, here is a blueprint for how we can achieve that :

inside the User.js file we need to add this function that would return the current status and (update it if the clear date has passed):

```js
function getUserStatus(userPersonalDetails) {
  const statusEmojiCode = lodashGet(
    userPersonalDetails,
    "status.emojiCode",
    ""
  );
  const statusText = lodashGet(userPersonalDetails, "status.text", "");
  const statusClearAfter = lodashGet(
    userPersonalDetails,
    "status.clearAfter",
    ""
  );

  if (statusClearAfter !== "" && new Date(statusClearAfter) < new Date()) {
    if (userPersonalDetails.accountID === currentUserAccountID) {
      clearCustomStatus();
      clearDraftCustomStatus();
    }
    return {
      emojiCode: "",
      text: "",
      clearAfter: "",
    };
  }
  return {
    emojiCode: statusEmojiCode,
    text: statusText,
    clearAfter: statusClearAfter,
  };
}
```

then we need to replace [this](https://github.com/Expensify/App/blob/541ee3568ba8034b85df295f8546a5d45f35f137/src/pages/settings/Profile/CustomStatus/StatusPage.js#L50):


with:

```js
const statusData = User.getUserStatus(currentUserPersonalDetails);
const currentUserEmojiCode = lodashGet(statusData, "emojiCode", "");
const currentUserStatusText = lodashGet(statusData, "text", "");
const currentUserClearAfter = lodashGet(statusData, "clearAfter", "");
```

same [here](https://github.com/Expensify/App/blob/addfa8860ac031bd4c3a98c563e124cc3a95154c/src/pages/settings/Profile/ProfilePage.js#L71): 

change it to:

```js
const status = User.getUserStatus(currentUserDetails);
const emojiCode = status.emojiCode;
```

and [here](https://github.com/Expensify/App/blob/25196becf259522dfe987a07bca8c31043bee22f/src/pages/home/report/ReportActionItemSingle.js#L88) change this: 

to:

```js
const { avatar, login, pendingFields, fallbackIcon } =
  personalDetails[actorAccountID] || {};
const status = User.getUserStatus(personalDetails[actorAccountID]);
```

and change [this](https://github.com/Expensify/App/blob/45494aef8ee06f05e07473388e07c2ed095ffc85/src/pages/home/sidebar/SignInOrAvatarWithOptionalStatus.js#L29):


to:

```js
const emojiStatus = User.getUserStatus(currentUserPersonalDetails).emojiCode;
```

and finally change [this](https://github.com/Expensify/App/blob/fa87f8bb98aa960f71584e188cbaee5c091bb0f0/src/components/LHNOptionsList/OptionRowLHN.js#L150-L152):



to

```js
const statusData = User.getUserStatus(optionItem);
const emojiCode = lodashGet(statusData, "emojiCode", "");
const statusText = lodashGet(statusData, "text", "");
const statusClearAfterDate = lodashGet(statusData, "clearAfter", "");
```

**OR**

alternatively we can make the above change on connecting to the App so it will update status if needed as we currently do with [the timezone](https://github.com/Expensify/App/blob/5d804d19d979f80f3bfdfbaf7e384d8e2f103ca4/src/libs/Navigation/AppNavigator/AuthScreens.tsx#L89-L105)

### Result

https://github.com/Expensify/App/assets/59809993/3f491858-e4d9-451d-aef5-82dcd9839054

**Comment by Pujan92** [comment]
Seems a backend issue where the status is not cleared after time expiration and the pusher event isn't receieved.

**Comment by stitesExpensify** [comment]
> Seems a backend issue where the status is not cleared after time expiration and the pusher event isn't receieved.


I agree

**Comment by stitesExpensify** [comment]
I think there are actually 2 issues here. One is that we aren't sending the time to the server in UTC (frontend bug), and the second is that the user who set the status does not get a pusher notification (backend bug).

When the bedrock job gets created on the server, it uses local time so if you are west of UTC it runs instantly and clears your status, and if you are east of UTC it happens late. 

Your status also disappears for other users, but not you.

**Comment by stitesExpensify** [comment]
@abzokhattab the server always assumes the time is in UTC so the frontend fix here is sending the datetime as UTC from the frontend as you mentioned here https://github.com/Expensify/App/issues/33128#issuecomment-1858637283

This will not fully fix the bug, but it will help

**Comment by stitesExpensify** [comment]
Can you write a formal proposal for that solution please @abzokhattab ?

**Comment by dylanexpensify** [comment]
thank you!

**Comment by abdulrahuman5196** [comment]
@abzokhattab Is the proposal ready as requested here? https://github.com/Expensify/App/issues/33128#issuecomment-1863353295

**Comment by dragnoir** [comment]
@stitesExpensify any idea how the BE handle the "clear after"? Does the BE process an empty emoji after or we just need to hide it on UI? 
I checked the Onyx values after the time of the clear period and it's not changed. I assume the BE is not doing any action!

**Comment by abdulrahuman5196** [comment]
> Looking more into it today

@abzokhattab Gentle ping

**Comment by dylanexpensify** [comment]
@abdulrahuman5196 to review ^! cc @deetergp

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by stitesExpensify** [comment]
@abzokhattab why do we need to convert in so many places? For things like `getEndOfToday` and `validateDateTimeIsAtLeastOneMinuteInFuture` can't we just use the local time for that, and then convert it right before we send it to the server?

**Comment by stitesExpensify** [comment]
> I checked the Onyx values after the time of the clear period and it's not changed. I assume the BE is not doing any action!

@dragnoir Right, that's the first problem I mentioned here https://github.com/Expensify/App/issues/33128#issuecomment-1863349962

**Comment by dylanexpensify** [comment]
@abzokhattab to reply

**Comment by abzokhattab** [comment]
> @abzokhattab why do we need to convert in so many places? For things like getEndOfToday and validateDateTimeIsAtLeastOneMinuteInFuture can't we just use the local time for that, and then convert it right before we send it to the server?

I wanted to make the dateutil function handle the conversion part themselves but this sounds like a lot of work and might cause breaking changes....


for that reason I am now considering the other way where we only change the clearAfter to UTC  on update..

I have updated [my proposal](https://github.com/Expensify/App/issues/33128#issuecomment-1858139232) please have a look

**Comment by stitesExpensify** [comment]
@abdulrahuman5196 can you take a look at the newest proposal please?

**Comment by abzokhattab** [comment]
@stitesExpensify  updated the proposal to use a middleware instead of parsing the date into a local date on the component level

**Comment by dylanexpensify** [comment]
@abdulrahuman5196 to confirm above

**Comment by dylanexpensify** [comment]
Bump @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
Sorry for missing this out.

**Comment by abdulrahuman5196** [comment]
> > I don't see any strong RCA mentioned for the timezone issue, more it seems like a hypothesis.
> 
> don't get your comment can elaborate more ... The RCA is that we don't convert the `clearAfter` to a UTC before it's sent to the backend it's rather sent in the client's local time not UTC.
> 
> Brandon mentioned that he will fix the original backend bug which is that the status is not cleared after the time has passed however he requested also that the date sent to the backend has to be in `UTC`

@abzokhattab  I understand your mention of we need to convert the time to UTC before sending it to the backend.

But why I mentioned it doesn't have strong RCA is because even if we did that, we can't verify if the change has fixed this issue or not without fixing the backend issues where the pusher notification is not sent as pointed out here - https://github.com/Expensify/App/issues/33128#issuecomment-1863349962

During my testing, I converted my timezone to UTC and tested, but the issue still persisted(I assume its due to push failure). The same would be happening at your testing as well, I assume that's why you had pointed out with the hypothesis here https://github.com/Expensify/App/issues/33128#issuecomment-1858637283.


So I wanted to confirm with @stitesExpensify on this requirement, since we won't be able to verify the fix if push notification is not sent? Correct me if wrong @stitesExpensify 
>Do we evaluate the proposal to just do the fix for `One is that we aren't sending the time to the server in UTC (frontend bug)` as pointed out here.

**Comment by stitesExpensify** [comment]
This should still be testable without fixing the backend issue. If you log in on one account in a normal tab and one account in an incognito tab, you will see the status expire correctly from the other user's point of view

**Comment by abdulrahuman5196** [comment]
Thanks for the info. Will check from other account

**Comment by dylanexpensify** [comment]
@abdulrahuman5196 any updates for us?

**Comment by abdulrahuman5196** [comment]
Caught in a priority issue today and yesterday. Will check and provide update tomorrow.

**Comment by abdulrahuman5196** [comment]
Still working on review. Will close out today.

**Comment by abdulrahuman5196** [comment]
I now tested by setting 2 account in UTC timezone and setting the status. Now the emoji status is cleared.

**Comment by abdulrahuman5196** [comment]
> I now tested by setting 2 account in UTC timezone and setting the status. Now the emoji status is cleared in the accounts.

@abzokhattab 's proposal here https://github.com/Expensify/App/issues/33128#issuecomment-1858139232 looks good and works well. My suggestion on the implementation options would be,

1) While setting/sending the time to API convert the time to UTC (Which means we will also store the time in UTC when we fetch the time from server)
2) While displaying the time, convert the UTC time to local time 
3) Avoid adding middlewares or disturbing current util functions we use, we can create/re-use different to change the UTC times. This should avoid the refractors mentioned in the proposal.

>And I have a curious question? What if user changes timezone after setting the clear status? Does backend handle it?
Anyways this is something I feel out of scope of this issue. Just putting out. @stitesExpensify 

🎀 👀 🎀
C+ Reviewed

cc: @stitesExpensify / @deetergp

**Comment by abdulrahuman5196** [comment]
> > Avoid adding middlewares or disturbing current util functions we use, we can create/re-use different to change the UTC times. This should avoid the refractors mentioned in the proposal.
> 
> i think the middleware is the most straightforward solution, can you provide more details about this part `we can create/re-use different to change the UTC times`

Sorry. Had a typo.
Updated.

>Avoid adding middlewares or disturbing current util functions we use, we can create/re-use different util functions to change the UTC times. This should avoid the refractors mentioned in the proposal.

My suggestion would be to not process the request/API through middleware. Let the time be stored in UTC timeformat in the onyx as well.

We can just convert the UTC time to local time while displaying in the view like you have suggested here
>we update the time to localdate in the component that use the status.

This makes a clear flow for time in the status segment. For all user interactions/display places we will convert the UTC to local time and in the ONYX /API flow we will use the UTC time.

**Comment by deetergp** [comment]
Sounds like we have a consensus on how to move forward. I that a datetime should only ever be local is when it is being displayed to the user, anywhere else, it ought to be in UTC.

**Comment by dylanexpensify** [comment]
@abzokhattab when can we get the PR raised?

**Comment by dylanexpensify** [comment]
Nice! @abdulrahuman5196 to get ready for review

**Comment by dylanexpensify** [comment]
bump @abzokhattab

**Comment by dylanexpensify** [comment]
bump @abdulrahuman5196

**Comment by abdulrahuman5196** [comment]
We are working on PR review. Still issues to solve.

**Comment by deetergp** [comment]
Any progress on that PR?

**Comment by abdulrahuman5196** [comment]
We are facing issues currently, Waiting on @stitesExpensify response for here https://github.com/Expensify/App/pull/35976#issuecomment-1976194287

**Comment by dylanexpensify** [comment]
@stitesExpensify any chance you've been able to tackle/ @abdulrahuman5196 ETA?

**Comment by stitesExpensify** [comment]
Not yet. This is held on me but it is not a priority at the moment unfortunately

**Comment by stitesExpensify** [comment]
@dylanexpensify I think that we should pay out for this issue. There was a lot of back and forth as well as development/reviewing time

**Comment by abzokhattab** [comment]
Kindly reminder @stitesExpensify @dylanexpensify

**Comment by dylanexpensify** [comment]
Apologies both! Yes, will do!

**Comment by dylanexpensify** [comment]
Payment summary:
- Contributor: @abzokhattab $500
- Contributor+: @abdulrahuman5196 $500
Please request/[apply](https://www.upwork.com/jobs/~011b7b18add80e8bb7)!

**Comment by abzokhattab** [comment]
Thank you @dylanexpensify i have just applied

**Comment by dylanexpensify** [comment]
offer sent @abzokhattab

**Comment by abdulrahuman5196** [comment]
@dylanexpensify Gentle ping on the above

**Comment by abdulrahuman5196** [comment]
Received payment
